From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dreeam <61569423+Dreeam-qwq@users.noreply.github.com>
Date: Thu, 24 Apr 2025 16:36:16 -0400
Subject: [PATCH] Paw optimization

Some random optimizations

- Remove Paper's dead code
- Only set shuffle random seed if is really used
- Cache direction values to skip copy

diff --git a/net/minecraft/data/advancements/packs/VanillaAdventureAdvancements.java b/net/minecraft/data/advancements/packs/VanillaAdventureAdvancements.java
index 4cb38f5e9fee280cc53bff5d9081c7a2a742835c..0f422b7a69c9e3710370deef10ec453624a231ff 100644
--- a/net/minecraft/data/advancements/packs/VanillaAdventureAdvancements.java
+++ b/net/minecraft/data/advancements/packs/VanillaAdventureAdvancements.java
@@ -1081,7 +1081,7 @@ public class VanillaAdventureAdvancements extends AdvancementSubProvider {
     }
 
     private Criterion<ItemUsedOnLocationTrigger.TriggerInstance> placedBlockActivatesCreakingHeart(final TagKey<Block> block) {
-        LootItemCondition.Builder[] conditions = Stream.of(Direction.values())
+        LootItemCondition.Builder[] conditions = Stream.of(Direction.VALUES) // Leaf - paw optimization - cache direction values
             .map(
                 direction -> {
                     StatePropertiesPredicate.Builder creakingHeartProperties = StatePropertiesPredicate.Builder.properties()
diff --git a/net/minecraft/data/loot/BlockLootSubProvider.java b/net/minecraft/data/loot/BlockLootSubProvider.java
index c9a6dcbd8c109029e31740753071100f155f4391..86a89f5b36be302cb026b6de52a6fcea22f98799 100644
--- a/net/minecraft/data/loot/BlockLootSubProvider.java
+++ b/net/minecraft/data/loot/BlockLootSubProvider.java
@@ -459,7 +459,7 @@ public abstract class BlockLootSubProvider implements LootTableSubProvider {
                             LootItem.lootTableItem(block)
                                 .when(condition)
                                 .apply(
-                                    Direction.values(),
+                                    Direction.VALUES, // Leaf - paw optimization - cache direction values
                                     dir -> SetItemCountFunction.setCount(ContextIntProviders.exactly(1), true)
                                         .when(
                                             MatchBlock.blockMatches(
@@ -484,7 +484,7 @@ public abstract class BlockLootSubProvider implements LootTableSubProvider {
                             block,
                             LootItem.lootTableItem(block)
                                 .apply(
-                                    Direction.values(),
+                                    Direction.VALUES, // Leaf - paw optimization - cache direction values
                                     dir -> SetItemCountFunction.setCount(ContextIntProviders.exactly(1), true)
                                         .when(
                                             MatchBlock.blockMatches(
diff --git a/net/minecraft/network/Connection.java b/net/minecraft/network/Connection.java
index 90010da198f4fd5ae34df7a3a85c548694bfdbf5..19dcc75746e56209296dd0ff560fde126e033ed9 100644
--- a/net/minecraft/network/Connection.java
+++ b/net/minecraft/network/Connection.java
@@ -658,13 +658,7 @@ public class Connection extends SimpleChannelInboundHandler<Packet<?>> {
             if (!(this.packetListener instanceof net.minecraft.server.network.ServerLoginPacketListenerImpl loginPacketListener)
                 || loginPacketListener.state != net.minecraft.server.network.ServerLoginPacketListenerImpl.State.VERIFYING
                 || Connection.joinAttemptsThisTick++ < MAX_PER_TICK) {
-            // Paper start - detailed watchdog information
-            net.minecraft.network.PacketProcessor.packetProcessing.push(this.packetListener);
-            try {
             tickable.tick();
-            } finally {
-                net.minecraft.network.PacketProcessor.packetProcessing.pop();
-            } // Paper end - detailed watchdog information
             } // Paper end - Buffer joins to world
         }
 
diff --git a/net/minecraft/server/level/ServerChunkCache.java b/net/minecraft/server/level/ServerChunkCache.java
index 00b58115fd29baead68820e7380c3b3777daf638..13bac576f49f34f2ef06545f33c6187d69d92369 100644
--- a/net/minecraft/server/level/ServerChunkCache.java
+++ b/net/minecraft/server/level/ServerChunkCache.java
@@ -591,8 +591,12 @@ public class ServerChunkCache extends ChunkSource implements ca.spottedleaf.moon
             if (org.dreeam.leaf.config.modules.opt.OptimizeMobSpawning.enabled) { this.level.natureSpawnChunkMap.tick(this.level, spawningChunks); } else { this.chunkMap.collectSpawningChunks(spawningChunks); } // Leaf - optimize mob spawning
             profiler.popPush("shuffleSpawningChunks");
             // Paper start - chunk tick iteration optimisation
+            // Leaf start - paw optimization - Only set seed if is really used
+            if (!this.level.paperConfig().entities.spawning.perPlayerMobSpawns) {
             this.shuffleRandom.setSeed(this.level.getRandom().nextLong());
-            if (!this.level.paperConfig().entities.spawning.perPlayerMobSpawns) Util.shuffle(spawningChunks, this.level.getRandom()); // Paper - Optional per player mob spawns; do not need this when per-player is enabled
+            Util.shuffle(spawningChunks, this.level.getRandom()); // Paper - Optional per player mob spawns; do not need this when per-player is enabled
+            }
+            // Leaf end - paw optimization - Only set seed if is really used
             // Paper end - chunk tick iteration optimisation
             profiler.popPush("tickSpawningChunks");
 
diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index a00c91040547f6fbd949b39b492b4f16eaf65f18..b11393cb9290e831105c94259d8c66defda2342f 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -1644,26 +1644,8 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
 
     }
 
-    // Paper start - log detailed entity tick information
-    // TODO replace with varhandle
-    static final java.util.concurrent.atomic.AtomicReference<Entity> currentlyTickingEntity = new java.util.concurrent.atomic.AtomicReference<>();
-
-    public static List<Entity> getCurrentlyTickingEntities() {
-        Entity ticking = currentlyTickingEntity.get();
-        List<Entity> ret = java.util.Arrays.asList(ticking == null ? me.titaniumtown.ArrayConstants.emptyEntityArray : new Entity[] { ticking }); // Gale - JettPack - reduce array allocations
-
-        return ret;
-    }
-    // Paper end - log detailed entity tick information
-
     public void tickNonPassenger(final Entity entity) {
-        // Paper start - log detailed entity tick information
         ca.spottedleaf.moonrise.common.util.TickThread.ensureTickThread("Cannot tick an entity off-main");
-        try {
-            if (currentlyTickingEntity.get() == null) {
-                currentlyTickingEntity.lazySet(entity);
-            }
-            // Paper end - log detailed entity tick information
         ProfilerFiller profiler = Profiler.get();
         profiler.push(entity.typeHolder()::getRegisteredName);
         profiler.incrementCounter("tickNonPassenger");
@@ -1678,13 +1660,6 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
         for (Entity passenger : entity.getPassengers()) {
             this.tickPassenger(entity, passenger, isActive); // Paper - EAR 2
         }
-        // Paper start - log detailed entity tick information
-        } finally {
-            if (currentlyTickingEntity.get() == entity) {
-                currentlyTickingEntity.lazySet(null);
-            }
-        }
-        // Paper end - log detailed entity tick information
     }
 
     private void tickPassenger(final Entity vehicle, final Entity entity, final boolean isActive) { // Paper - EAR 2
diff --git a/net/minecraft/util/ParticleUtils.java b/net/minecraft/util/ParticleUtils.java
index 995f2336dc9c10b871d92db912a6ef584f5d8b8c..f2ac7bd5c584a6533bab0fb027eaef394b3d69bd 100644
--- a/net/minecraft/util/ParticleUtils.java
+++ b/net/minecraft/util/ParticleUtils.java
@@ -19,7 +19,7 @@ public class ParticleUtils {
     ) {
         RandomSource random = level.getRandom();
 
-        for (Direction direction : Direction.values()) {
+        for (Direction direction : Direction.VALUES) { // Leaf - paw optimization - cache direction values
             spawnParticlesOnBlockFace(level, pos, particle, particlesPerFaceRange, direction, () -> getRandomSpeedRanges(random), 0.55);
         }
     }
diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index b41c3707bcb0c8685033d434b01a522e5cb4db25..70d7e9db4d453cba15e478ebfe192070b804ecb9 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -1202,13 +1202,6 @@ public abstract class Entity
     // Paper start - detailed watchdog information
     public final Object posLock = new Object(); // Paper - log detailed entity tick information
 
-    @Nullable
-    private Vec3 moveVector;
-    private double moveStartX;
-    private double moveStartY;
-    private double moveStartZ;
-    // Paper end - detailed watchdog information
-
     public void move(final MoverType moverType, Vec3 delta) {
         if (SharedConstants.IS_RUNNING_IN_IDE && !moverType.isServerAndClientSimulated() && !this.canSimulateMovement()) {
             throw (IllegalStateException)Util.pauseInIde(
@@ -1217,16 +1210,7 @@ public abstract class Entity
         }
 
         final Vec3 originalMovement = delta; // Paper - Expose pre-collision velocity
-        // Paper start - detailed watchdog information
         ca.spottedleaf.moonrise.common.util.TickThread.ensureTickThread("Cannot move an entity off-main");
-        synchronized (this.posLock) {
-            this.moveStartX = this.getX();
-            this.moveStartY = this.getY();
-            this.moveStartZ = this.getZ();
-            this.moveVector = delta;
-        }
-        try {
-        // Paper end - detailed watchdog information
         if (this.noPhysics) {
             this.setPos(this.getX() + delta.x, this.getY() + delta.y, this.getZ() + delta.z);
             this.recordMovement(moverType, delta);
@@ -1367,13 +1351,6 @@ public abstract class Entity
                 profiler.pop();
             }
         }
-        // Paper start - detailed watchdog information
-        } finally {
-            synchronized (this.posLock) { // Paper
-                this.moveVector = null;
-            } // Paper
-        }
-        // Paper end - detailed watchdog information
     }
 
     public void recordMovement(final MoverType moverType, final Vec3 movement) {
@@ -5215,9 +5192,7 @@ public abstract class Entity
 
     public void setDeltaMovement(final Vec3 deltaMovement) {
         if (deltaMovement.isFinite()) {
-            synchronized (this.posLock) { // Paper - detailed watchdog information
             this.deltaMovement = deltaMovement;
-            } // Paper - detailed watchdog information
         }
     }
 
@@ -5322,9 +5297,7 @@ public abstract class Entity
         }
         // Paper end - Block invalid positions and bounding box
         if (this.position.x != x || this.position.y != y || this.position.z != z) {
-            synchronized (this.posLock) { // Paper - detailed watchdog information
             this.position = new Vec3(x, y, z);
-            } // Paper - detailed watchdog information
             int fx = Mth.floor(x);
             int fy = Mth.floor(y);
             int fz = Mth.floor(z);
diff --git a/net/minecraft/world/entity/monster/Ghast.java b/net/minecraft/world/entity/monster/Ghast.java
index c5aa0c1a6fa8849525b79200f8fb5a915d6543ca..9c8efe9eba76ba601d2aa440461e7eb6b6a1e663 100644
--- a/net/minecraft/world/entity/monster/Ghast.java
+++ b/net/minecraft/world/entity/monster/Ghast.java
@@ -531,7 +531,7 @@ public class Ghast extends Mob implements Enemy {
                 return false;
             }
 
-            for (Direction dir : Direction.values()) {
+            for (Direction dir : Direction.VALUES) { // Leaf - paw optimization - cache direction values
                 for (int i = 1; i < distanceToBlocks; i++) {
                     BlockPos offset = pos.relative(dir, i);
                     if (!level.getBlockState(offset).isAir()) {
diff --git a/net/minecraft/world/entity/monster/Shulker.java b/net/minecraft/world/entity/monster/Shulker.java
index d12c44eab3068685ac8e6d630a4b7b8ff2b8a5e1..4c578c6890b636eab674bb1f18100b14beb10d10 100644
--- a/net/minecraft/world/entity/monster/Shulker.java
+++ b/net/minecraft/world/entity/monster/Shulker.java
@@ -414,7 +414,7 @@ public class Shulker extends AbstractGolem implements Enemy {
     }
 
     protected @Nullable Direction findAttachableSurface(final BlockPos target) {
-        for (Direction direction : Direction.values()) {
+        for (Direction direction : Direction.VALUES) { // Leaf - paw optimization - cache direction values
             if (this.canStayAt(target, direction)) {
                 return direction;
             }
diff --git a/net/minecraft/world/level/block/BaseCoralPlantTypeBlock.java b/net/minecraft/world/level/block/BaseCoralPlantTypeBlock.java
index ba7bc939b5eab5a182fb9ca161e4800bb1bd4a1d..fe8cc752d2f3e85f5ecd19ac06d182b0f2dca2e3 100644
--- a/net/minecraft/world/level/block/BaseCoralPlantTypeBlock.java
+++ b/net/minecraft/world/level/block/BaseCoralPlantTypeBlock.java
@@ -42,7 +42,7 @@ public abstract class BaseCoralPlantTypeBlock extends Block implements SimpleWat
             return true;
         }
 
-        for (Direction direction : Direction.values()) {
+        for (Direction direction : Direction.VALUES) { // Leaf - paw optimization - cache direction values
             if (level.getFluidState(blockPos.relative(direction)).is(FluidTags.WATER)) {
                 return true;
             }
diff --git a/net/minecraft/world/level/block/BaseFireBlock.java b/net/minecraft/world/level/block/BaseFireBlock.java
index 580fac7e521e635fa1177652a2e6a287585589d2..022b7436e6a181b6a193cba8d81b4cbdb9714505 100644
--- a/net/minecraft/world/level/block/BaseFireBlock.java
+++ b/net/minecraft/world/level/block/BaseFireBlock.java
@@ -211,7 +211,7 @@ public abstract class BaseFireBlock extends Block {
 
         BlockPos.MutableBlockPos testPos = pos.mutable();
 
-        for (Direction face : Direction.values()) {
+        for (Direction face : Direction.VALUES) { // Leaf - paw optimization - cache direction values
             BlockState state = level.getBlockState(testPos.set(pos).move(face));
             if (PortalShape.FRAME.test(state)) {
                 Direction.Axis preferredAxis = forwardDirection.getAxis().isHorizontal()
diff --git a/net/minecraft/world/level/block/ConcretePowderBlock.java b/net/minecraft/world/level/block/ConcretePowderBlock.java
index cfefd61ff179238f061c66f44acb18eafbf30055..5bc6862a2175a587892b705ba5ea42edd1926817 100644
--- a/net/minecraft/world/level/block/ConcretePowderBlock.java
+++ b/net/minecraft/world/level/block/ConcretePowderBlock.java
@@ -58,7 +58,7 @@ public class ConcretePowderBlock extends FallingBlock {
         boolean touchesLiquid = false;
         BlockPos.MutableBlockPos testPos = pos.mutable();
 
-        for (Direction direction : Direction.values()) {
+        for (Direction direction : Direction.VALUES) { // Leaf - paw optimization - cache direction values
             BlockState blockState = level.getBlockState(testPos);
             if (direction != Direction.DOWN || canSolidify(blockState)) {
                 testPos.setWithOffset(pos, direction);
diff --git a/net/minecraft/world/level/block/CoralBlock.java b/net/minecraft/world/level/block/CoralBlock.java
index f4bbc483ed8c4ffc3b994a0528cb60fbc4737403..cf78e2887e59882e763a1c5b14a9421777574fbc 100644
--- a/net/minecraft/world/level/block/CoralBlock.java
+++ b/net/minecraft/world/level/block/CoralBlock.java
@@ -54,7 +54,7 @@ public class CoralBlock extends Block {
 
     protected boolean scanForWater(final BlockGetter level, final BlockPos blockPos) {
         if (!((net.minecraft.world.level.LevelAccessor) level).getMinecraftWorld().purpurConfig.coralDieOutsideWater) return true; // Purpur - Config to not let coral die
-        for (Direction direction : Direction.values()) {
+        for (Direction direction : Direction.VALUES) { // Leaf - paw optimization - cache direction values
             FluidState fluidState = level.getFluidState(blockPos.relative(direction));
             if (fluidState.is(FluidTags.WATER)) {
                 return true;
diff --git a/net/minecraft/world/level/block/CreakingHeartBlock.java b/net/minecraft/world/level/block/CreakingHeartBlock.java
index 63ddfd24b55f98409dbaea02bc0eb0c9604a0400..9e590da924caabc764e035491315ba8482e6232d 100644
--- a/net/minecraft/world/level/block/CreakingHeartBlock.java
+++ b/net/minecraft/world/level/block/CreakingHeartBlock.java
@@ -120,7 +120,7 @@ public class CreakingHeartBlock extends BaseEntityBlock {
     }
 
     private static boolean isSurroundedByLogs(final LevelAccessor level, final BlockPos pos) {
-        for (Direction dir : Direction.values()) {
+        for (Direction dir : Direction.VALUES) { // Leaf - paw optimization - cache direction values
             BlockPos neighbourPos = pos.relative(dir);
             BlockState neighbourState = level.getBlockState(neighbourPos);
             if (!neighbourState.is(BlockTags.PALE_OAK_LOGS)) {
diff --git a/net/minecraft/world/level/block/DiodeBlock.java b/net/minecraft/world/level/block/DiodeBlock.java
index fc007b164c7fc4e9608d02c64c5cd4dc6934b4e2..ed8539eb5ff73596335f34e0f9e1d49caac54999 100644
--- a/net/minecraft/world/level/block/DiodeBlock.java
+++ b/net/minecraft/world/level/block/DiodeBlock.java
@@ -85,7 +85,7 @@ public abstract class DiodeBlock extends HorizontalDirectionalBlock {
                 dropResources(state, level, pos, blockEntity);
                 level.removeBlock(pos, false);
 
-                for (Direction direction : Direction.values()) {
+                for (Direction direction : Direction.VALUES) { // Leaf - paw optimization - cache direction values
                     level.updateNeighborsAt(pos.relative(direction), this);
                 }
             }
diff --git a/net/minecraft/world/level/block/FireBlock.java b/net/minecraft/world/level/block/FireBlock.java
index 132edd0e94456378cc8598c39e20749b3cfb382a..f5fac3e66529a9342830e770ae46ac6cb2400454 100644
--- a/net/minecraft/world/level/block/FireBlock.java
+++ b/net/minecraft/world/level/block/FireBlock.java
@@ -129,7 +129,7 @@ public class FireBlock extends BaseFireBlock {
         if (!this.canBurn(belowState) && !belowState.isFaceSturdy(level, below, Direction.UP)) {
             BlockState result = this.defaultBlockState();
 
-            for (Direction direction : Direction.values()) {
+            for (Direction direction : Direction.VALUES) { // Leaf - paw optimization - cache direction values
                 BooleanProperty property = PROPERTY_BY_DIRECTION.get(direction);
                 if (property != null) {
                     result = result.setValue(property, this.canBurn(level.getBlockState(pos.relative(direction))));
@@ -299,7 +299,7 @@ public class FireBlock extends BaseFireBlock {
     }
 
     private boolean isValidFireLocation(final BlockGetter level, final BlockPos pos) {
-        for (Direction direction : Direction.values()) {
+        for (Direction direction : Direction.VALUES) { // Leaf - paw optimization - cache direction values
             if (this.canBurn(level.getBlockState(pos.relative(direction)))) {
                 return true;
             }
@@ -315,7 +315,7 @@ public class FireBlock extends BaseFireBlock {
 
         int odds = 0;
 
-        for (Direction direction : Direction.values()) {
+        for (Direction direction : Direction.VALUES) { // Leaf - paw optimization - cache direction values
             BlockState blockState = level.getBlockState(pos.relative(direction));
             odds = Math.max(this.getIgniteOdds(blockState), odds);
         }
diff --git a/net/minecraft/world/level/block/FrostedIceBlock.java b/net/minecraft/world/level/block/FrostedIceBlock.java
index e0cf2165a1459636bf1c7a41234e5906b61e63ac..6f9aa6d3b1fd88aea27bd15cdc2675c1e922a13d 100644
--- a/net/minecraft/world/level/block/FrostedIceBlock.java
+++ b/net/minecraft/world/level/block/FrostedIceBlock.java
@@ -42,7 +42,7 @@ public class FrostedIceBlock extends IceBlock {
             if (brightness > 11 - state.getValue(AGE) - state.getLightDampening() && this.slightlyMelt(state, level, pos)) {
                 BlockPos.MutableBlockPos neighborPos = new BlockPos.MutableBlockPos();
 
-                for (Direction direction : Direction.values()) {
+                for (Direction direction : Direction.VALUES) { // Leaf - paw optimization - cache direction values
                     neighborPos.setWithOffset(pos, direction);
                     BlockState neighbour = level.getBlockState(neighborPos);
                     if (neighbour.is(this) && !this.slightlyMelt(neighbour, level, neighborPos)) {
@@ -88,7 +88,7 @@ public class FrostedIceBlock extends IceBlock {
         int result = 0;
         BlockPos.MutableBlockPos neighborPos = new BlockPos.MutableBlockPos();
 
-        for (Direction direction : Direction.values()) {
+        for (Direction direction : Direction.VALUES) { // Leaf - paw optimization - cache direction values
             neighborPos.setWithOffset(pos, direction);
             if (level.getBlockState(neighborPos).is(this)) {
                 if (++result >= limit) {
diff --git a/net/minecraft/world/level/block/LeavesBlock.java b/net/minecraft/world/level/block/LeavesBlock.java
index 186492b2264d9042a534ce6e71a1686d7e773390..fbd3b99d13bf4bc874f82e88d3d586f23b7403cf 100644
--- a/net/minecraft/world/level/block/LeavesBlock.java
+++ b/net/minecraft/world/level/block/LeavesBlock.java
@@ -124,7 +124,7 @@ public class LeavesBlock extends Block implements SimpleWaterloggedBlock {
         int newDistance = 7;
         BlockPos.MutableBlockPos neighborPos = new BlockPos.MutableBlockPos();
 
-        for (Direction direction : Direction.values()) {
+        for (Direction direction : Direction.VALUES) { // Leaf - paw optimization - cache direction values
             neighborPos.setWithOffset(pos, direction);
             newDistance = Math.min(newDistance, getDistanceAt(level.getBlockState(neighborPos)) + 1);
             if (newDistance == 1) {
diff --git a/net/minecraft/world/level/block/MultifaceBlock.java b/net/minecraft/world/level/block/MultifaceBlock.java
index b03ad91538137b9d04bad6431672d1fe865496ff..09f2da4fc7a8e7770a3b4fd259048245ddc2cea3 100644
--- a/net/minecraft/world/level/block/MultifaceBlock.java
+++ b/net/minecraft/world/level/block/MultifaceBlock.java
@@ -67,7 +67,7 @@ public class MultifaceBlock extends Block implements SimpleWaterloggedBlock {
 
         Set<Direction> faces = EnumSet.noneOf(Direction.class);
 
-        for (Direction direction : Direction.values()) {
+        for (Direction direction : Direction.VALUES) { // Leaf - paw optimization - cache direction values
             if (hasFace(state, direction)) {
                 faces.add(direction);
             }
@@ -79,7 +79,7 @@ public class MultifaceBlock extends Block implements SimpleWaterloggedBlock {
     public static Set<Direction> unpack(final byte data) {
         Set<Direction> presentDirections = EnumSet.noneOf(Direction.class);
 
-        for (Direction direction : Direction.values()) {
+        for (Direction direction : Direction.VALUES) { // Leaf - paw optimization - cache direction values
             if ((data & (byte)(1 << direction.ordinal())) > 0) {
                 presentDirections.add(direction);
             }
diff --git a/net/minecraft/world/level/block/RedStoneOreBlock.java b/net/minecraft/world/level/block/RedStoneOreBlock.java
index e3d997672f73df19dce794ee4f81e1b9c830f387..204a2ec8b03a3a49e776094243431a894ad0545b 100644
--- a/net/minecraft/world/level/block/RedStoneOreBlock.java
+++ b/net/minecraft/world/level/block/RedStoneOreBlock.java
@@ -132,7 +132,7 @@ public class RedStoneOreBlock extends Block {
         double offset = 0.5625;
         RandomSource random = level.getRandom();
 
-        for (Direction direction : Direction.values()) {
+        for (Direction direction : Direction.VALUES) { // Leaf - paw optimization - cache direction values
             BlockPos relative = pos.relative(direction);
             if (!level.getBlockState(relative).isSolidRender()) {
                 Direction.Axis axis = direction.getAxis();
diff --git a/net/minecraft/world/level/block/RedstoneTorchBlock.java b/net/minecraft/world/level/block/RedstoneTorchBlock.java
index e6680b9bfa001e6c00598256f13d4e74b64dc32c..6a652a6cca52dccbc1e244eea6c26ffab4517f12 100644
--- a/net/minecraft/world/level/block/RedstoneTorchBlock.java
+++ b/net/minecraft/world/level/block/RedstoneTorchBlock.java
@@ -42,7 +42,7 @@ public class RedstoneTorchBlock extends BaseTorchBlock {
     private void notifyNeighbors(final Level level, final BlockPos pos, final BlockState state) {
         Orientation orientation = this.randomOrientation(level, state);
 
-        for (Direction direction : Direction.values()) {
+        for (Direction direction : Direction.VALUES) { // Leaf - paw optimization - cache direction values
             level.updateNeighborsAt(pos.relative(direction), this, ExperimentalRedstoneUtils.withFront(orientation, direction));
         }
     }
diff --git a/net/minecraft/world/level/block/RedstoneWireBlock.java b/net/minecraft/world/level/block/RedstoneWireBlock.java
index 2afcc3ee9438bd09eec452e28bc5dc92e4a74927..78be705043916e80a542ca0991a58f55780c6776 100644
--- a/net/minecraft/world/level/block/RedstoneWireBlock.java
+++ b/net/minecraft/world/level/block/RedstoneWireBlock.java
@@ -359,7 +359,7 @@ public class RedstoneWireBlock extends Block {
         if (level.getBlockState(pos).is(this)) {
             level.updateNeighborsAt(pos, this);
 
-            for (Direction direction : Direction.values()) {
+            for (Direction direction : Direction.VALUES) { // Leaf - paw optimization - cache direction values
                 level.updateNeighborsAt(pos.relative(direction), this);
             }
         }
@@ -387,7 +387,7 @@ public class RedstoneWireBlock extends Block {
     @Override
     protected void affectNeighborsAfterRemoval(final BlockState state, final ServerLevel level, final BlockPos pos, final boolean movedByPiston) {
         if (!movedByPiston) {
-            for (Direction direction : Direction.values()) {
+            for (Direction direction : Direction.VALUES) { // Leaf - paw optimization - cache direction values
                 level.updateNeighborsAt(pos.relative(direction), this);
             }
 
diff --git a/net/minecraft/world/level/block/SculkSensorBlock.java b/net/minecraft/world/level/block/SculkSensorBlock.java
index 6e217dea06365d2125b51b8222f49e186fcf6c08..78145c3c30e818a60a4d5ae920b14ed0075a7f90 100644
--- a/net/minecraft/world/level/block/SculkSensorBlock.java
+++ b/net/minecraft/world/level/block/SculkSensorBlock.java
@@ -250,7 +250,7 @@ public class SculkSensorBlock extends BaseEntityBlock implements SimpleWaterlogg
     }
 
     public static void tryResonateVibration(final @Nullable Entity sourceEntity, final Level level, final BlockPos pos, final int vibrationFrequency) {
-        for (Direction direction : Direction.values()) {
+        for (Direction direction : Direction.VALUES) { // Leaf - paw optimization - cache direction values
             BlockPos relativePos = pos.relative(direction);
             BlockState blockState = level.getBlockState(relativePos);
             if (blockState.is(BlockTags.VIBRATION_RESONATORS)) {
diff --git a/net/minecraft/world/level/block/entity/CreakingHeartBlockEntity.java b/net/minecraft/world/level/block/entity/CreakingHeartBlockEntity.java
index 9272f75f939eeccb0b163bd858446c371cfc52ff..a8d24a5eea0353fde49221948fe396f0c9d6f4f2 100644
--- a/net/minecraft/world/level/block/entity/CreakingHeartBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/CreakingHeartBlockEntity.java
@@ -253,7 +253,7 @@ public class CreakingHeartBlockEntity extends BlockEntity {
         RandomSource random = level.getRandom();
         Mutable<BlockPos> placedResin = new MutableObject<>(null);
         BlockPos.breadthFirstTraversal(this.worldPosition, 2, 64, (pos, acceptor) -> {
-            for (Direction dir : Util.shuffledCopy(Direction.values(), random)) {
+            for (Direction dir : Util.shuffledCopy(Direction.VALUES, random)) { // Leaf - paw optimization - cache direction values
                 BlockPos neighbourPos = pos.relative(dir);
                 if (level.getBlockState(neighbourPos).is(BlockTags.PALE_OAK_LOGS)) {
                     acceptor.accept(neighbourPos);
@@ -264,7 +264,7 @@ public class CreakingHeartBlockEntity extends BlockEntity {
                 return BlockPos.TraversalNodeStatus.ACCEPT;
             }
 
-            for (Direction dir : Util.shuffledCopy(Direction.values(), random)) {
+            for (Direction dir : Util.shuffledCopy(Direction.VALUES, random)) { // Leaf - paw optimization - cache direction values
                 BlockPos neightbourPos = pos.relative(dir);
                 BlockState neighbourState = level.getBlockState(neightbourPos);
                 Direction opposite = dir.getOpposite();
diff --git a/net/minecraft/world/level/block/entity/TheEndGatewayBlockEntity.java b/net/minecraft/world/level/block/entity/TheEndGatewayBlockEntity.java
index fa06068346351cd7d78c7560a0f09ab1282886c1..26c73790480ab90f0e90046fd3cc3ea42b12782a 100644
--- a/net/minecraft/world/level/block/entity/TheEndGatewayBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/TheEndGatewayBlockEntity.java
@@ -262,7 +262,7 @@ public class TheEndGatewayBlockEntity extends TheEndPortalBlockEntity {
     public int getParticleAmount() {
         int count = 0;
 
-        for (Direction direction : Direction.values()) {
+        for (Direction direction : Direction.VALUES) { // Leaf - paw optimization - cache direction values
             count += this.shouldRenderFace(direction) ? 1 : 0;
         }
 
diff --git a/net/minecraft/world/level/block/piston/PistonBaseBlock.java b/net/minecraft/world/level/block/piston/PistonBaseBlock.java
index 92afbd653c6aaff973110d24357f68140e75d418..954dc301346863fc1d421f6f96660a6de55f17e7 100644
--- a/net/minecraft/world/level/block/piston/PistonBaseBlock.java
+++ b/net/minecraft/world/level/block/piston/PistonBaseBlock.java
@@ -115,7 +115,7 @@ public class PistonBaseBlock extends DirectionalBlock {
     }
 
     private boolean getNeighborSignal(final SignalGetter level, final BlockPos pos, final Direction pushDirection) {
-        for (Direction direction : Direction.values()) {
+        for (Direction direction : Direction.VALUES) { // Leaf - paw optimization - cache direction values
             if (direction != pushDirection && level.hasSignal(pos.relative(direction), direction)) {
                 return true;
             }
@@ -127,7 +127,7 @@ public class PistonBaseBlock extends DirectionalBlock {
 
         BlockPos above = pos.above();
 
-        for (Direction direction : Direction.values()) {
+        for (Direction direction : Direction.VALUES) { // Leaf - paw optimization - cache direction values
             if (direction != Direction.DOWN && level.hasSignal(above.relative(direction), direction)) {
                 return true;
             }
diff --git a/net/minecraft/world/level/block/piston/PistonStructureResolver.java b/net/minecraft/world/level/block/piston/PistonStructureResolver.java
index cc9e9bfbb500c44aa5d5e7ae02fc77cdef8ace29..89cd9de3952743dc4c05d6b2df89a64da8944331 100644
--- a/net/minecraft/world/level/block/piston/PistonStructureResolver.java
+++ b/net/minecraft/world/level/block/piston/PistonStructureResolver.java
@@ -175,7 +175,7 @@ public class PistonStructureResolver {
     private boolean addBranchingBlocks(final BlockPos fromPos) {
         BlockState fromState = this.level.getBlockState(fromPos);
 
-        for (Direction direction : Direction.values()) {
+        for (Direction direction : Direction.VALUES) { // Leaf - paw optimization - cache direction values
             if (direction.getAxis() != this.pushDirection.getAxis()) {
                 BlockPos neighbourPos = fromPos.relative(direction);
                 BlockState neighbourState = this.level.getBlockState(neighbourPos);
diff --git a/net/minecraft/world/level/block/state/pattern/BlockPattern.java b/net/minecraft/world/level/block/state/pattern/BlockPattern.java
index ed5a5fb38bfe8fb9216c7eab00b833485c4bbade..c305bc4a4988faddd29fb2cee11920fb2d546660 100644
--- a/net/minecraft/world/level/block/state/pattern/BlockPattern.java
+++ b/net/minecraft/world/level/block/state/pattern/BlockPattern.java
@@ -78,8 +78,8 @@ public class BlockPattern {
         int dist = Math.max(Math.max(this.width, this.height), this.depth);
 
         for (BlockPos testPos : BlockPos.betweenClosed(origin, origin.offset(dist - 1, dist - 1, dist - 1))) {
-            for (Direction forwards : Direction.values()) {
-                for (Direction up : Direction.values()) {
+            for (Direction forwards : Direction.VALUES) { // Leaf - paw optimization - cache direction values
+                for (Direction up : Direction.VALUES) { // Leaf - paw optimization - cache direction values
                     if (up != forwards && up != forwards.getOpposite()) {
                         BlockPattern.BlockPatternMatch match = this.matches(testPos, forwards, up, cache);
                         if (match != null) {
diff --git a/net/minecraft/world/level/chunk/UpgradeData.java b/net/minecraft/world/level/chunk/UpgradeData.java
index 431a48073ad71ed10aa423a0d261e83378f904be..2496bd51bc814963a6be386512717d3b7caa5ada 100644
--- a/net/minecraft/world/level/chunk/UpgradeData.java
+++ b/net/minecraft/world/level/chunk/UpgradeData.java
@@ -146,14 +146,13 @@ public class UpgradeData {
             int maxX = chunkPos.getMinBlockX() + (!singular || !north && !south ? (west ? 0 : 15) : 14);
             int minZ = chunkPos.getMinBlockZ() + (!singular || !east && !west ? (north ? 0 : 15) : 1);
             int maxZ = chunkPos.getMinBlockZ() + (!singular || !east && !west ? (north ? 0 : 15) : 14);
-            Direction[] updateDirections = Direction.values();
             BlockPos.MutableBlockPos neighbourPos = new BlockPos.MutableBlockPos();
 
             for (BlockPos pos : BlockPos.betweenClosed(minX, level.getMinY(), minZ, maxX, level.getMaxY(), maxZ)) {
                 BlockState state = level.getBlockState(pos);
                 BlockState newState = state;
 
-                for (Direction direction : updateDirections) {
+                for (Direction direction : Direction.VALUES) { // Leaf - paw optimization - cache direction values
                     neighbourPos.setWithOffset(pos, direction);
                     newState = updateState(newState, direction, level, pos, neighbourPos);
                 }
@@ -181,7 +180,6 @@ public class UpgradeData {
             int[] upgradeIndex = this.index[sectionIndex];
             this.index[sectionIndex] = null;
             if (upgradeIndex != null && upgradeIndex.length > 0) {
-                Direction[] directions = Direction.values();
                 PalettedContainer<BlockState> states = chunkSection.getStates();
                 int sectionY = chunk.getSectionYFromSectionIndex(sectionIndex);
                 int bottomYInSection = SectionPos.sectionToBlockCoord(sectionY);
@@ -194,7 +192,7 @@ public class UpgradeData {
                     BlockState state = states.get(coordinate);
                     BlockState newState = state;
 
-                    for (Direction direction : directions) {
+                    for (Direction direction : Direction.VALUES) { // Leaf - paw optimization - cache direction values
                         neighbourPos.setWithOffset(pos, direction);
                         if (SectionPos.blockToSectionCoord(pos.getX()) == chunkPos.x() && SectionPos.blockToSectionCoord(pos.getZ()) == chunkPos.z()) {
                             newState = updateState(newState, direction, level, pos, neighbourPos);
diff --git a/net/minecraft/world/level/gameevent/vibrations/VibrationSystem.java b/net/minecraft/world/level/gameevent/vibrations/VibrationSystem.java
index 7364d6ae99772912bf353f7a004dd26f00fa80ae..98e52c705e04f56e77f679eb05f3e0b8df4be743 100644
--- a/net/minecraft/world/level/gameevent/vibrations/VibrationSystem.java
+++ b/net/minecraft/world/level/gameevent/vibrations/VibrationSystem.java
@@ -271,7 +271,7 @@ public interface VibrationSystem {
             Vec3 from = new Vec3(Mth.floor(origin.x) + 0.5, Mth.floor(origin.y) + 0.5, Mth.floor(origin.z) + 0.5);
             Vec3 to = new Vec3(Mth.floor(dest.x) + 0.5, Mth.floor(dest.y) + 0.5, Mth.floor(dest.z) + 0.5);
 
-            for (Direction direction : Direction.values()) {
+            for (Direction direction : Direction.VALUES) { // Leaf - paw optimization - cache direction values
                 Vec3 nudgedSource = from.relative(direction, 1.0E-5F);
                 if (level.isBlockInLine(new ClipBlockStateContext(nudgedSource, to, state -> state.is(BlockTags.OCCLUDES_VIBRATION_SIGNALS))).getType()
                     != HitResult.Type.BLOCK) {
diff --git a/net/minecraft/world/level/levelgen/feature/AbstractOreFeature.java b/net/minecraft/world/level/levelgen/feature/AbstractOreFeature.java
index 304e5d98b3367480eca22380af684f2cd22dd907..9c4129c6ae9a2ed998586669bb85215d53f560e1 100644
--- a/net/minecraft/world/level/levelgen/feature/AbstractOreFeature.java
+++ b/net/minecraft/world/level/levelgen/feature/AbstractOreFeature.java
@@ -68,7 +68,7 @@ public abstract class AbstractOreFeature implements Feature {
     public static boolean checkNeighbors(final Function<BlockPos, BlockState> blockGetter, final BlockPos pos, final Predicate<BlockState> predicate) {
         BlockPos.MutableBlockPos neighborPos = new BlockPos.MutableBlockPos();
 
-        for (Direction direction : Direction.values()) {
+        for (Direction direction : Direction.VALUES) { // Leaf - paw optimization - cache direction values
             neighborPos.setWithOffset(pos, direction);
             if (predicate.test(blockGetter.apply(neighborPos))) {
                 return true;
diff --git a/net/minecraft/world/level/levelgen/feature/BlueIceFeature.java b/net/minecraft/world/level/levelgen/feature/BlueIceFeature.java
index 1d516c0a7449d9d4b905d71b7b19343eb6a4d3ba..a29263420512852f4b842596e91e22b050c2e262 100644
--- a/net/minecraft/world/level/levelgen/feature/BlueIceFeature.java
+++ b/net/minecraft/world/level/levelgen/feature/BlueIceFeature.java
@@ -30,7 +30,7 @@ public record BlueIceFeature() implements Feature {
 
         boolean foundPackedIce = false;
 
-        for (Direction direction : Direction.values()) {
+        for (Direction direction : Direction.VALUES) { // Leaf - paw optimization - cache direction values
             if (direction != Direction.DOWN && level.getBlockState(origin.relative(direction)).is(Blocks.PACKED_ICE)) {
                 foundPackedIce = true;
                 break;
@@ -54,7 +54,7 @@ public record BlueIceFeature() implements Feature {
                 BlockPos placePos = origin.offset(random.nextInt(xzDiff) - random.nextInt(xzDiff), yOff, random.nextInt(xzDiff) - random.nextInt(xzDiff));
                 BlockState placeState = level.getBlockState(placePos);
                 if (placeState.isAir() || placeState.is(Blocks.WATER) || placeState.is(Blocks.PACKED_ICE) || placeState.is(Blocks.ICE)) {
-                    for (Direction direction : Direction.values()) {
+                    for (Direction direction : Direction.VALUES) { // Leaf - paw optimization - cache direction values
                         BlockState relativeBlockState = level.getBlockState(placePos.relative(direction));
                         if (relativeBlockState.is(Blocks.BLUE_ICE)) {
                             level.setBlock(placePos, Blocks.BLUE_ICE.defaultBlockState(), Block.UPDATE_CLIENTS);
diff --git a/net/minecraft/world/level/levelgen/feature/TreeFeature.java b/net/minecraft/world/level/levelgen/feature/TreeFeature.java
index 92eaf1c49070276c5eeae37625440c6034489f19..1920819aa5c9b82eb3a16d61fe3613a63fa2d982 100644
--- a/net/minecraft/world/level/levelgen/feature/TreeFeature.java
+++ b/net/minecraft/world/level/levelgen/feature/TreeFeature.java
@@ -233,7 +233,7 @@ public record TreeFeature(
 
                     shape.fill(pos.getX() - bounds.minX(), pos.getY() - bounds.minY(), pos.getZ() - bounds.minZ());
 
-                    for (Direction direction : Direction.values()) {
+                    for (Direction direction : Direction.VALUES) { // Leaf - paw optimization - cache direction values
                         neighborPos.setWithOffset(pos, direction);
                         if (bounds.isInside(neighborPos)) {
                             int xInShape = neighborPos.getX() - bounds.minX();
diff --git a/net/minecraft/world/level/levelgen/feature/VinesFeature.java b/net/minecraft/world/level/levelgen/feature/VinesFeature.java
index 3721ed5894df54e250c1e34a612ffe0b903956c3..700a10f0194d86a29faca13dd29c96ddaa010037 100644
--- a/net/minecraft/world/level/levelgen/feature/VinesFeature.java
+++ b/net/minecraft/world/level/levelgen/feature/VinesFeature.java
@@ -24,7 +24,7 @@ public record VinesFeature() implements Feature {
             return false;
         }
 
-        for (Direction direction : Direction.values()) {
+        for (Direction direction : Direction.VALUES) { // Leaf - paw optimization - cache direction values
             if (direction != Direction.DOWN && VineBlock.isAcceptableNeighbour(level, origin.relative(direction), direction)) {
                 level.setBlock(origin, Blocks.VINE.defaultBlockState().setValue(VineBlock.getPropertyForFace(direction), true), Block.UPDATE_CLIENTS);
                 return true;
diff --git a/net/minecraft/world/level/levelgen/feature/treedecorators/CreakingHeartDecorator.java b/net/minecraft/world/level/levelgen/feature/treedecorators/CreakingHeartDecorator.java
index 9828a63f8451f8f6e2fe39c2c4efe30a4aef8045..1e1e62ba9f009e785995a4ad48f8a3ecb5b7a036 100644
--- a/net/minecraft/world/level/levelgen/feature/treedecorators/CreakingHeartDecorator.java
+++ b/net/minecraft/world/level/levelgen/feature/treedecorators/CreakingHeartDecorator.java
@@ -38,7 +38,7 @@ public class CreakingHeartDecorator extends TreeDecorator {
                 List<BlockPos> heartPlacements = new ArrayList<>(logs);
                 Util.shuffle(heartPlacements, random);
                 Optional<BlockPos> targetPos = heartPlacements.stream().filter(pos -> {
-                    for (Direction dir : Direction.values()) {
+                    for (Direction dir : Direction.VALUES) { // Leaf - paw optimization - cache direction values
                         if (!context.checkBlock(pos.relative(dir), state -> state.is(BlockTags.LOGS))) {
                             return false;
                         }
diff --git a/net/minecraft/world/level/levelgen/structure/structures/BuriedTreasurePieces.java b/net/minecraft/world/level/levelgen/structure/structures/BuriedTreasurePieces.java
index 7c9704dfbde05d1075e1a7b6f66162dd0679a592..56ef89f0e3fea4b09825e61cf44c29e4ccf201d4 100644
--- a/net/minecraft/world/level/levelgen/structure/structures/BuriedTreasurePieces.java
+++ b/net/minecraft/world/level/levelgen/structure/structures/BuriedTreasurePieces.java
@@ -54,7 +54,7 @@ public class BuriedTreasurePieces {
                     || belowState.is(Blocks.DIORITE)) {
                     BlockState softState = !currentState.isAir() && !isLiquid(currentState) ? currentState : Blocks.SAND.defaultBlockState();
 
-                    for (Direction direction : Direction.values()) {
+                    for (Direction direction : Direction.VALUES) { // Leaf - paw optimization - cache direction values
                         BlockPos relativePos = pos.relative(direction);
                         BlockState relativeState = level.getBlockState(relativePos);
                         if (relativeState.isAir() || isLiquid(relativeState)) {
diff --git a/net/minecraft/world/level/levelgen/structure/structures/DesertPyramidStructure.java b/net/minecraft/world/level/levelgen/structure/structures/DesertPyramidStructure.java
index 8b2f75cfa141681f1b01af336b8ad19c58131c71..5b775d09a94ef6e7125485b72128b5043b23f20f 100644
--- a/net/minecraft/world/level/levelgen/structure/structures/DesertPyramidStructure.java
+++ b/net/minecraft/world/level/levelgen/structure/structures/DesertPyramidStructure.java
@@ -49,7 +49,7 @@ public class DesertPyramidStructure extends SinglePieceStructure {
             }
         }
 
-        ObjectArrayList<BlockPos> shuffledSandPlacements = new ObjectArrayList<>(uniqueSandPlacements.stream().toList());
+        ObjectArrayList<BlockPos> shuffledSandPlacements = new ObjectArrayList<>(uniqueSandPlacements); // Leaf - paw optimization - TODO: use array
         RandomSource positionalRandom = RandomSource.createThreadLocalInstance(level.getSeed()).forkPositional().at(pieces.calculateBoundingBox().getCenter());
         Util.shuffle(shuffledSandPlacements, positionalRandom);
         int suspiciousSandToPlace = Math.min(uniqueSandPlacements.size(), positionalRandom.nextInt(5, 8));
diff --git a/net/minecraft/world/level/levelgen/structure/structures/MineshaftPieces.java b/net/minecraft/world/level/levelgen/structure/structures/MineshaftPieces.java
index a4dc64b0325b30ca2d4b710013a803f26dbbd9ad..62b0112011bd74247c5fd55a61645f0dd6a5e4b4 100644
--- a/net/minecraft/world/level/levelgen/structure/structures/MineshaftPieces.java
+++ b/net/minecraft/world/level/levelgen/structure/structures/MineshaftPieces.java
@@ -605,7 +605,7 @@ public class MineshaftPieces {
             BlockPos.MutableBlockPos worldPos = this.getWorldPos(x, y, z);
             int sturdyNeighbours = 0;
 
-            for (Direction direction : Direction.values()) {
+            for (Direction direction : Direction.VALUES) { // Leaf - paw optimization - cache direction values
                 worldPos.move(direction);
                 if (chunkBB.isInside(worldPos) && level.getBlockState(worldPos).isFaceSturdy(level, worldPos, direction.getOpposite())) {
                     if (++sturdyNeighbours >= count) {
diff --git a/net/minecraft/world/level/levelgen/structure/structures/OceanMonumentPieces.java b/net/minecraft/world/level/levelgen/structure/structures/OceanMonumentPieces.java
index fa93d0f67d439547c37fe3ae5ca95ad195850a15..2b646f0a8fafbf32364cba3bda9a7162f81646cc 100644
--- a/net/minecraft/world/level/levelgen/structure/structures/OceanMonumentPieces.java
+++ b/net/minecraft/world/level/levelgen/structure/structures/OceanMonumentPieces.java
@@ -258,7 +258,7 @@ public class OceanMonumentPieces {
                     for (int y = 0; y < 3; y++) {
                         int pos = getRoomIndex(x, y, z);
                         if (roomGrid[pos] != null) {
-                            for (Direction direction : Direction.values()) {
+                            for (Direction direction : Direction.VALUES) { // Leaf - paw optimization - cache direction values
                                 int neighX = x + direction.getStepX();
                                 int neighY = y + direction.getStepY();
                                 int neighZ = z + direction.getStepZ();
diff --git a/net/minecraft/world/level/material/LavaFluid.java b/net/minecraft/world/level/material/LavaFluid.java
index 2608fcdff409e1f342157cfc11042e0b2f592a4a..15973912d617a036d802e37bd6a9ae710e3904a3 100644
--- a/net/minecraft/world/level/material/LavaFluid.java
+++ b/net/minecraft/world/level/material/LavaFluid.java
@@ -141,7 +141,7 @@ public abstract class LavaFluid extends FlowingFluid {
     }
 
     private boolean hasFlammableNeighbours(final LevelReader level, final BlockPos pos) {
-        for (Direction direction : Direction.values()) {
+        for (Direction direction : Direction.VALUES) { // Leaf - paw optimization - cache direction values
             if (this.isFlammable(level, pos.relative(direction))) {
                 return true;
             }
diff --git a/net/minecraft/world/level/pathfinder/AmphibiousNodeEvaluator.java b/net/minecraft/world/level/pathfinder/AmphibiousNodeEvaluator.java
index 21d969b49e22c00f456befc7bea25a927de6f2fb..745cc35d9591003908a0c571ffecbe2506adbd82 100644
--- a/net/minecraft/world/level/pathfinder/AmphibiousNodeEvaluator.java
+++ b/net/minecraft/world/level/pathfinder/AmphibiousNodeEvaluator.java
@@ -97,7 +97,7 @@ public class AmphibiousNodeEvaluator extends WalkNodeEvaluator {
         if (blockPathType == PathType.WATER) {
             BlockPos.MutableBlockPos reusablePos = new BlockPos.MutableBlockPos();
 
-            for (Direction direction : Direction.values()) {
+            for (Direction direction : Direction.VALUES) { // Leaf - paw optimization - cache direction values
                 reusablePos.set(x, y, z).move(direction);
                 PathType pathType = context.getPathTypeFromState(reusablePos.getX(), reusablePos.getY(), reusablePos.getZ());
                 if (pathType == PathType.BLOCKED) {
diff --git a/net/minecraft/world/level/pathfinder/SwimNodeEvaluator.java b/net/minecraft/world/level/pathfinder/SwimNodeEvaluator.java
index 355142be30daaa00809cd77e99857dc812c3d791..49d959160337f61a40cb94c31aec8dad0aee3839 100644
--- a/net/minecraft/world/level/pathfinder/SwimNodeEvaluator.java
+++ b/net/minecraft/world/level/pathfinder/SwimNodeEvaluator.java
@@ -51,7 +51,7 @@ public class SwimNodeEvaluator extends NodeEvaluator {
         int count = 0;
         Map<Direction, Node> nodes = Maps.newEnumMap(Direction.class);
 
-        for (Direction direction : Direction.values()) {
+        for (Direction direction : Direction.VALUES) { // Leaf - paw optimization - cache direction values
             Node node = this.findAcceptedNode(pos.x + direction.getStepX(), pos.y + direction.getStepY(), pos.z + direction.getStepZ());
             nodes.put(direction, node);
             if (this.isNodeValid(node)) {
diff --git a/net/minecraft/world/level/redstone/DefaultRedstoneWireEvaluator.java b/net/minecraft/world/level/redstone/DefaultRedstoneWireEvaluator.java
index 9027dfa971e3d3b83d4647691e3332f1e851c760..82ee509ffe0d3be94ae9b7580f60f1ac3c04f697 100644
--- a/net/minecraft/world/level/redstone/DefaultRedstoneWireEvaluator.java
+++ b/net/minecraft/world/level/redstone/DefaultRedstoneWireEvaluator.java
@@ -34,7 +34,7 @@ public class DefaultRedstoneWireEvaluator extends RedstoneWireEvaluator {
             Set<BlockPos> toUpdate = Sets.newHashSet();
             toUpdate.add(pos);
 
-            for (Direction direction : Direction.values()) {
+            for (Direction direction : Direction.VALUES) { // Leaf - paw optimization - cache direction values
                 toUpdate.add(pos.relative(direction));
             }
 
diff --git a/net/minecraft/world/level/redstone/Orientation.java b/net/minecraft/world/level/redstone/Orientation.java
index f71dc76e61103a80d5d8d574bfb0648fef3f5abc..dd76e72daa7fb382806acf166c75a1815d73a6aa 100644
--- a/net/minecraft/world/level/redstone/Orientation.java
+++ b/net/minecraft/world/level/redstone/Orientation.java
@@ -136,7 +136,7 @@ public class Orientation {
             self.withSideBias.put(sideBias, generateContext(new Orientation(self.up, self.front, sideBias), lookup));
         }
 
-        for (Direction facing : Direction.values()) {
+        for (Direction facing : Direction.VALUES) { // Leaf - paw optimization - cache direction values
             Direction up = self.up;
             if (facing == self.up) {
                 up = self.front.getOpposite();
@@ -149,7 +149,7 @@ public class Orientation {
             self.withFront.put(facing, generateContext(new Orientation(up, facing, self.sideBias), lookup));
         }
 
-        for (Direction facing : Direction.values()) {
+        for (Direction facing : Direction.VALUES) { // Leaf - paw optimization - cache direction values
             Direction front = self.front;
             if (facing == self.front) {
                 front = self.up.getOpposite();
